taar/recommenders/collaborative_recommender.py

```diff
@@ -103,51 +120,52 @@ def can_recommend(self, client_data, extra_data={}):
-        # space.
-        distances = {}
-        for addon in self.raw_item_matrix:
-            # We don't really need to show the items we requested.
-            # They will always end up with the greatest score. Also
-            # filter out legacy addons from the suggestions.
-            hashed_id = addon.get("id")
-            str_hashed_id = str(hashed_id)
-            if (
-                hashed_id in installed_addons_as_hashes
-                or str_hashed_id not in self.addon_mapping
-                or self.addon_mapping[str_hashed_id].get("isWebextension", False)
-                is False
-            ):
-                continue
-
-            dist = np.dot(user_factors_transposed, addon.get("features"))
-            # Read the addon ids from the "addon_mapping" looking it
-            # up by 'id' (which is an hashed value).
-            addon_id = self.addon_mapping[str_hashed_id].get("id")
-            distances[addon_id] = dist
-
-        # Sort the suggested addons by their score and return the
-        # sorted list of addon ids.
-        sorted_dists = sorted(distances.items(), key=op.itemgetter(1), reverse=True)
-        recommendations = [(s[0], s[1]) for s in sorted_dists[:limit]]
+            # Build the query vector by setting the position of the queried addons to 1.0
+            # and the other to 0.0.
+            query_vector = np.array(
+                [
+                    1.0 if (entry.get("id") in installed_addons_as_hashes) else 0.0
+                    for entry in self.raw_item_matrix
+                ]
+            )
+
+            # Build the user factors matrix.
+            user_factors = np.matmul(query_vector, self.model)
+            user_factors_transposed = np.transpose(user_factors)
+
+            # Compute the distance between the user and all the addons in the latent
+            # space.
+            distances = {}
+            for addon in self.raw_item_matrix:
+                # We don't really need to show the items we requested.
+                # They will always end up with the greatest score. Also
+                # filter out legacy addons from the suggestions.
+                hashed_id = addon.get("id")
+                str_hashed_id = str(hashed_id)
+                if (
+                    hashed_id in installed_addons_as_hashes
+                    or str_hashed_id not in self.addon_mapping
+                    or self.addon_mapping[str_hashed_id].get("isWebextension", False)
+                    is False
+                ):
+                    continue
+
+                dist = np.dot(user_factors_transposed, addon.get("features"))
+                # Read the addon ids from the "addon_mapping" looking it
+                # up by 'id' (which is an hashed value).
+                addon_id = self.addon_mapping[str_hashed_id].get("id")
+                distances[addon_id] = dist
+
+            # Sort the suggested addons by their score and return the
```

Comment:
Thread safety isn't need when we're just logging and returning the final computed value.